components/raftstore/src/store/async_io.rs

```diff
@@ -0,0 +1,949 @@
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+{
+    pub fn new(router: RaftRouter<EK, ER>) -> Self {
+        Self { router }
+    }
+}
+
+impl<EK, ER> Notifier for RegionNotifier<EK, ER>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+{
+    fn notify_persisted(
+        &self,
+        region_id: u64,
+        peer_id: u64,
+        ready_number: u64,
+        send_time: Instant,
+    ) {
+        if let Err(e) = self.router.force_send(
+            region_id,
+            PeerMsg::Persisted {
+                peer_id,
+                ready_number,
+                send_time,
+            },
+        ) {
+            warn!(
+                "failed to send noop to trigger persisted ready";
+                "region_id" => region_id,
+                "peer_id" => peer_id,
+                "ready_number" => ready_number,
+                "error" => ?e,
+            );
+        }
+    }
+
+    fn notify_unreachable(&self, region_id: u64, to_peer_id: u64) {
+        let msg = SignificantMsg::Unreachable {
+            region_id,
+            to_peer_id,
+        };
+        if let Err(e) = self
+            .router
+            .force_send(region_id, PeerMsg::SignificantMsg(msg))
+        {
+            warn!(
+                "failed to send unreachable";
+                "region_id" => region_id,
+                "unreachable_peer_id" => to_peer_id,
+                "error" => ?e,
+            );
+        }
+    }
+}
+
+/// AsyncWriteTask contains write tasks which need to be persisted to kv db and raft db.
+#[derive(Debug)]
+pub struct AsyncWriteTask<EK, ER>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+{
+    region_id: u64,
+    peer_id: u64,
+    ready_number: u64,
+    pub send_time: Instant,
+    pub kv_wb: Option<EK::WriteBatch>,
+    pub raft_wb: Option<ER::LogBatch>,
+    pub entries: Vec<Entry>,
+    pub cut_logs: Option<(u64, u64)>,
+    pub raft_state: Option<RaftLocalState>,
+    pub messages: Vec<RaftMessage>,
+    pub proposal_times: Vec<Instant>,
+}
+
+impl<EK, ER> AsyncWriteTask<EK, ER>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+{
+    pub fn new(region_id: u64, peer_id: u64, ready_number: u64) -> Self {
+        Self {
+            region_id,
+            peer_id,
+            ready_number,
+            send_time: Instant::now(),
+            kv_wb: None,
+            raft_wb: None,
+            entries: vec![],
+            cut_logs: None,
+            raft_state: None,
+            messages: vec![],
+            proposal_times: vec![],
+        }
+    }
+
+    pub fn has_data(&self) -> bool {
+        !(self.raft_state.is_none()
+            && self.entries.is_empty()
+            && self.cut_logs.is_none()
+            && self.kv_wb.as_ref().map_or(true, |wb| wb.is_empty())
+            && self.raft_wb.as_ref().map_or(true, |wb| wb.is_empty()))
+    }
+
+    /// Sanity check for robustness.
+    pub fn valid(&self) -> Result<()> {
+        if self.region_id == 0 || self.peer_id == 0 || self.ready_number == 0 {
```

Comment:
Why not `assert` or `debug_assert`? I think it can't recover from this error.

Reply:
The caller will panic if this function returns an error.